server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -161,9 +163,9 @@ public BytesReference readBytesReference(int length) throws IOException {
         if (length == 0) {
             return BytesArray.EMPTY;
         }
-        byte[] bytes = new byte[length];
-        readBytes(bytes, 0, length);
-        return new BytesArray(bytes, 0, length);
+        final ByteArray byteArray = BigArrays.NON_RECYCLING_INSTANCE.newByteArray(length);
+        byteArray.set(0, this, length);
```

Comment:
Could we achieve the same goal without adding another method to ByteArray? ByteArray doesn't currently know anything about StreamInput (other than a static helper method that is only used by tests...).

Ultimately we need a BytesReference here. I think that can be better achieved by specializing this method for StreamInput implementations. For example, for BytesReferenceStreamInput we should just call slice?

I don't like adding a StreamInput method to ByteArray. Rather I think this should be inverted so the construction/copying happens within StreamInput. This might still require adding a ByteArray method, but could be more abstract?

@original-brownbear do you have any thoughts here?

Reply:
++ on inverting it and adding a method on StreamInput instead, and leaving to ByteArray just the necessary (abstract) logic.

Reply:
By the way, my original intention was to use a plain java.io.InputStream here no a Elasticsearch StreamInput. 

A more abstract way would be to introduce a functional interface that fills a provided byte array like:

```
void set(long index, ReadStream);

public interface ReadStream {
  /**
   * Reads {@code length} bytes on the provided byte[] at the given offset.
   */
  void readBytes(byte[] bytes, int offset, int length);
}
```
Is any of those options apealing?